### PR TITLE
Fix remote use of gef.memory.maps

### PR DESCRIPTION
**Pull request by gordonmessmer**
## Description

This change passes both the "display" path and the local "real" path for each section or file to the print_got_for function in GotCommand.  Without this change, the `got` command does not work properly in a remote debugging session.

## Checklist

-  [x] My code follows the code style of this project.
-  [x] My change includes a change to the documentation, if required.
-  [x] If my change adds new code, [adequate tests](docs/testing.md) have been added.
-  [x] I have read and agree to the **CONTRIBUTING** document.

### CODE REVIEW

**Comment by gordonmessmer** [comment]
One way to make this commit a little nicer would be to add an interface that both Sections and the GefSessionManager.file property could implement, each providing both a display path (.path) and a local path (.realpath) attribute/property.

**Comment by gordonmessmer** [comment]
I've rebased this commit on the new main branch.

Currently, this patch allows the `got` command without `--all` to work, and it allows shared objects whose paths are canonical to work with using the `--all` option, but non-canonical paths are silently ignored.  So, it's partially functional, but not really elegant.

It could be merged as-is.  If you have any suggestions on making it better fit the code style of the project, I can work on those.  I could also look for a way to print a warning for files that can't be found (which will probably always be non-canonical paths).  At a stretch, I could even try to add some simple searches for likely alternate paths when the realpath isn't present.

**Comment by hugsy** [comment]
No problem. `--all` is about to be merged anyway.
Also I usually never review draft PR, so when you deem it ready, mark it as such as you can also assign me for review so I get a notification.

Cheers

**Comment by gordonmessmer** [comment]
I added a patch that performs one permutation to find paths that aren't found at the expected path.  With that patch, remote debugging against a Red Hat system provides the full expected output for `got --all`.

**Comment by gordonmessmer** [comment]
(I did not remove the note about the bug from the got documentation, because shared objects that are not found due to *other* symlinks could still result in their absence from the output.)

**Comment by gordonmessmer** [comment]
Another way to make this implementation less messy would be to always use maps, rather than using maps for `--all` and gef.session.file/get_filepath otherwise.

I don't have strong feelings about patch 3, but it's an option.  I could squash that and the first, or drop it.

**Comment by gordonmessmer** [comment]
"you can also assign me for review so I get a notification"

@hugsy as a non-project member, I don't think I can.  :laughing:

File: gef.py
Comment (Grazfather) [inline gef.py:699]: Slightly better to do
```python
candidate1 = pathlib.Path(gef.session.remote.root) / self.path
```
Since it doesn't require building strings and doesn't use os specific path separators.

File: gef.py
Comment (Grazfather) [inline gef.py:702]: Can we just check for a symlink?

File: gef.py
Comment (Grazfather) [inline gef.py:705]: better to use trim instead of hardcoding the number of characters.

Comment (Grazfather) [reply gef.py:705]: Same idea with building the path here.

File: gef.py
Comment (Grazfather) [inline gef.py:9318]: `(bool(args.all) or mapfile.path == str(gef.session.file)) and ...`

File: gef.py
Comment (Grazfather) [inline gef.py:693]: Make this a doc string

File: gef.py
Comment (gordonmessmer) [reply gef.py:699]: Can't do that, because self.path is an absolute path, which causes the initial path to be ignored.

Still, might not be idiomatic.  Could use .join() if that's preferred.

File: gef.py
Comment (gordonmessmer) [reply gef.py:702]: Sadly, no, for the same reason that we can't use realpath(), which is the suggested solution upstream.  The symlinks don't exist on the host where gdb is running.  They aren't copied by the remote debugging session.

File: gef.py
Comment (gordonmessmer) [reply gef.py:705]: Trim?

I can certainly use `len('/usr')` if that's preferred.

File: gef.py
Comment (gordonmessmer) [reply gef.py:9318]: Yeah, I realized later that I'd made that more complex than necessary.

File: gef.py
Comment (gordonmessmer) [reply gef.py:693]: Good call

File: gef.py
Comment (gordonmessmer) [reply gef.py]: Updated this section using pathlib more.

File: gef.py
Comment (gordonmessmer) [reply gef.py]: Updated this using pathlib, too.

File: gef.py
Comment (hugsy) [inline gef.py:701]: you need to check if the session is remote before using `gef.session.remote` otherwise, it will throw an exception

File: gef.py
Comment (hugsy) [inline gef.py:709]: Wouldn't this check always be true?

File: gef.py
Comment (hugsy) [inline gef.py:716]: Wouldn't make more sense to make `realpath` a `Path`? Avoiding all the Path <-> str conversions?

**Review by hugsy — CHANGES_REQUESTED** [review]
It would be nice to add a test case on this function specifically (i.e. `search_for_realpath`)

File: gef.py
Comment (gordonmessmer) [reply gef.py:701]: Right now, search_for_realpath() is only called after checking if the session is remote (line 718).  Would you like an assert here?

File: gef.py
Comment (gordonmessmer) [reply gef.py:709]: No, I don't think so.  This check modifies the path in a way that might be true for one type of remote system, but it's just a guess.  It's not guaranteed to match a path after modifying it.

File: gef.py
Comment (gordonmessmer) [reply gef.py:716]: If we change the type, we have to at least review all of the internal uses, and all of the uses in gef-extras, and we also risk breaking any third-party extensions.  Changing the type would be a breaking change to the API, so I assume that's not something desirable.

But that's your call to make.  Let me know what you'd like to see, here.

File: gef.py
Comment (gordonmessmer) [reply gef.py:701]: (Or just mark the function private, so that it doesn't gain new users which haven't already checked gef.session.remote?)

File: gef.py
Comment (hugsy) [reply gef.py:716]: Fair point, maybe something to work in the future then.

**Comment by gordonmessmer** [comment]
So I think the outstanding work is test for search_for_realpath, and make that function private (and assert gef.session.remote).

If I resolved anything by mistake, please let me know.

**Comment by gordonmessmer** [comment]
So... This is strange, to me:

I'm adding a test to tests/api/gef_memory.py::GefMemoryApi, based on test_func_parse_maps_remote_gdbserver.  While that process is attached, there are sections for `/tmp/default.out`, `/usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2`, `[stack]`, `[vvar]`, `[vdso]`, and `[vsyscall]`, but *not* for `/usr/lib/x86_64-linux-gnu/libc.so.6`, which it is definitely linked to.

Do either of you have any guess why that might be?

**Comment by gordonmessmer** [comment]
I'll have to think about this some more... The test currently checked in succeeds on ubuntu 22.04, but not 20.04.

I *could* just not assert that a lib in /usr was found, but that opens the possibility of some future test platform not exhibiting this behavior, and the test doing nothing at all.

**Comment by gordonmessmer** [comment]
What's important to you in these tests?

Both Debian and Fedora family systems have the same issue that I originally noticed with remote debugging: /lib and /lib64 are symlinks into /usr, and the files retrieved by gdb haven't been resolved by realpath(), so they don't match the paths indicated in /proc/pid/maps.  That makes testing easier, because I don't need to introduce tests on a different platform, nor to I need to modify the local tmp root to reproduce the issue.

But Debian *also* makes ld.so a symlink, and that makes testing on Debian(/Ubuntu) very difficult, because fixing up that path isn't simple and predictable the way that fixing the /usr symlinks is, so remote binaries on Debian systems always have a map that's missing, which looks like a failure in the realpath property of map sections.

On both Ubuntu 20.04 and 22.04, the maps file will contain only the binary executable, ld.so, and virtual objects, typically.  But if I add `gdb.execute("b main"); gdb.execute("continue")` before checking maps, I'll get a copy of maps after runtime linking is complete on Ubuntu 22.04 -- but not on Ubuntu 20.04.  That's making testing difficult, because one platform doesn't behave like the other for reasons that aren't obvious to me.

I think the right way to do test the realpath property would be an isolated unit test, not a functional test, but gef isn't designed to be imported for unit testing.

These tests pass, but only because on Ubuntu 20.04 they don't do anything.

**Comment by gordonmessmer** [comment]
Checking in:  Let me know if you'd like any more work on this PR.

**Comment by hugsy** [comment]
> Checking in: Let me know if you'd like any more work on this PR.

Hi @gordonmessmer !

I haven't forgotten, but I cannot easily review/test your PR.
I'll try do that on my return next week.

**Comment by gordonmessmer** [comment]
NP.  I just like to check in on open PRs every week or two to make sure that I know if someone is waiting on me.  :)

**Comment by Grazfather** [comment]
Could you come up with step to repro the issue so that I can test it? Let's get this merged :)

**Comment by gordonmessmer** [comment]
The shortest reproduction case is probably simply:

In one terminal, run `gdbserver :9999 /bin/bash`

In a second terminal, run `gdb` with GEF enabled, and then run `got --all`.

You'll see the error:

```
[!] Command 'got' failed to execute properly, reason: unsupported format string passed to GefRemoteSessionManager.__format__
```

That's because the Section.realpath property is broken, and attempts to use `f"/tmp/gef/{gef.session.remote:d}/{self.path}"` instead of something like `f"{gef.session.remote.root}/{self.path}"`.  But if you fix that and run `got --all`, you'll still see that the GOT is printed for the main binary but not for any libraries, because no libraries can be found in the remote session root, because the paths in /proc/pid/maps are canonical, but the paths returned by the remote gdbserver are not.

**Comment by Grazfather** [comment]
Interestingly, while trying to repro I couldn't get even the format string error. I was sourcing an old work tree, but that means that this is some sort of regression.

On main I can repro and on your branch things mostly look good, but I only get output for the binary e.g. the same output as when I don't use `--all`

**Comment by gordonmessmer** [comment]
If you don't get any additional libraries, then most likely the paths provided by gdbserver do not match the canonical path for some reason other than a /usr prefix.

Looking at this again, I see that the suffix is *also* likely to be not canonical due to minor-version sonames, so I'll have to add more code to prune numeric suffixes during the realpath search.

On your system, take a look at /proc/pid/maps for the process you're debugging, and compare the paths that appear in that file to the paths that appear in the gdb session remote root to see if your system has issues other than symlinks to `/usr/lib*` or the numeric suffixes.

I'll send another patch asap, but if you see something not covered yet, we'd need to expand the search function further.

**Comment by Grazfather** [comment]
I am testing with a local gdb server so shouldn’t they match?

**Comment by gordonmessmer** [comment]
Paths will often not match due to the bug we discussed in https://github.com/hugsy/gef/pull/1101#issuecomment-2132702144

**Comment by Grazfather** [comment]
Thanks. I’ll take a look and try to debug when I’m at a computer.

**Comment by gordonmessmer** [comment]
I've added another function which should improve the realpath search success rate.

I'd like to believe that this is good enough for now, fixing the vast majority of cases.  It's definitely significantly better than the status quo.  If more symlinks are identified in the future that the realpath search should test, the function can always be expanded.

**Comment by Grazfather** [comment]
Part of my issue was that I was testing on `_start`, and I had to get to proper main for libc to get loaded and downloaded.

It's looking good now!

File: gef.py
Comment (Grazfather) [reply gef.py:716]: An assert will probably let us detect bugs and give better errors if we ever hit that situation.

Comment (Grazfather) [reply gef.py:716]: You could also make it so that you pass in `gef.remote` as an argument.

**Review by Grazfather — APPROVED** [review]

File: gef.py
Comment (gordonmessmer) [reply gef.py:717]: I've added an assert.

Passing gef.session.remote as an argument to a private function doesn't make much sense to me, since the realpath property that calls the search function doesn't take arguments.

**Comment by mahaloz** [comment]
Any way this is related to #1070 that went stale? The library loading part is what makes me think so.

**Comment by gordonmessmer** [comment]
> Any way this is related to https://github.com/hugsy/gef/pull/1070 that went stale?

Yes and no.

Yes: If the issue that 1070 was intended to fix were fixed, then the tests introduced by this PR would (I think) operate identically on Ubuntu 20.04 and 22.04.  And if that happened, then the tests could *require* that some library realpath was fixed in a remote session.  That would be an improvement.  And it's possible that this PR will not improve the status quo on Ubuntu 20.04 because the maps file is simply out of date and doesn't refresh.

But otherwise, no.  There's minimal relation between the two PRs.  Neither PR is a substitute for the other.

File: gef.py
Comment (hugsy) [reply gef.py:717]: FTR I'm very happy with the `assert`, like graz said "if" it fails we'll know immediately why

File: tests/api/gef_memory.py
Comment (hugsy) [inline tests/api/gef_memory.py:177]: `is True` is almost never necessary, but not a blocker

File: gef.py
Comment (hugsy) [inline gef.py:692]: Wrong return type
```suggestion
    def _search_for_realpath_without_versions(self, path: pathlib.Path) -> Optional[str]:
```

File: gef.py
Comment (hugsy) [inline gef.py:707]: Same
```suggestion
    def _search_for_realpath(self) -> Optional[str]:
```

File: gef.py
Comment (hugsy) [inline gef.py:718]: `candidate1` is never re-used, could be omitted because it brings little, and the comment explains better the purpose of the test
```suggestion
        if (gef.session.remote.root / remote_path.relative_to(remote_path.anchor)).is_file():
```

**Review by hugsy — COMMENTED** [review]
Tests are green all across, so LGTM. Just a few minor code quality comments

File: gef.py
Comment (gordonmessmer) [reply gef.py:718]: It's re-used in the return on the following line.

Comment (gordonmessmer) [reply gef.py:718]: (And also in the subsequent call to _search_for_realpath_without_versions()

**Comment by gordonmessmer** [comment]
Thank you for the review.  Feedback should be addressed in the new commit.

File: gef.py
Comment (hugsy) [reply gef.py:718]: Ah ok, disregard, I misunderstood the function

**Comment by hugsy** [comment]
> Thank you for the review. Feedback should be addressed in the new commit.

Sweet, I'll merge right after then! Thanks @gordonmessmer

**Comment by gordonmessmer** [comment]
Sorry, I meant https://github.com/hugsy/gef/pull/1109/commits/6e5a96dfef5f41ebba7d09005c825c9e0f370fab

**Review by hugsy — APPROVED** [review]

**Comment by hugsy** [comment]
All merged, great work again @gordonmessmer 🎉

**Pull request merged**
